vowpalwabbit/offset_tree.cc

```diff
@@ -186,20 +199,65 @@ namespace VW { namespace offset_tree {
 
     return scores;
   }
-  
-  void learn(offset_tree& tree, single_learner& base, example& ec)
+
+void offset_tree::learn(LEARNER::single_learner& base, example& ec) {
+  const auto global_action = ec.l.cb.costs[0].action;
+  const auto global_weight = ec.weight;
+
+  // for brevity
+  auto& nodes = binary_tree.nodes; 
+
+  tree_node node = nodes[global_action - 1];
+  do
+  {  // ascend
+    const auto previous_id = node.id;
+    node = nodes[node.parent_id];
+
+    // learn
+    uint32_t local_action = 2;
+    if (node.left_id == previous_id)
+      local_action = 1;
+    ec.l.cb.costs[0].action = local_action;
+    base.learn(ec, node.id - binary_tree.leaf_node_count());
+
+    // re-weight
+    base.predict(ec, node.id - binary_tree.leaf_node_count());
+    ec.weight *= ec.pred.a_s[local_action - 1].score;
+  } while (node.parent_id != node.id);
+
+  ec.l.cb.costs[0].action = global_action;
+  ec.weight = global_weight;
+}
+
+inline void copy_to_action_scores(const offset_tree::scores_t& scores, ACTION_SCORE::action_scores& a_s) {
+  a_s.clear();
+  for (uint32_t idx = 0; idx < scores.size(); ++idx)
   {
-    THROW("Offset tree learn() - not yet impemented.");
+    a_s.push_back({idx, scores[idx]});
   }
+}
 
-  void predict(offset_tree& ot, single_learner& base, example& ec)
+void predict(offset_tree& ot, single_learner& base, example& ec)
+{
+  // get predictions for all internal nodes in binary tree.
+  ec.pred.a_s.clear();
+  const auto& scores = ot.predict(base, ec);
+  copy_to_action_scores(scores, ec.pred.a_s);
```

Comment:
The pattern where we have unnecessary copies seems undesirable from LOC and efficiency perspectives.   Maybe instead using the pattern where pass in a return reference is better?